### PR TITLE
Adaption for ONT analysis

### PR DESCRIPTION
**Pull request by YalanBi**
adaption for ONT analysis
gene structure variation quantification
triangle plot visualisation
gene entropy calculation
export transcript sequences
some small fixes

### CODE REVIEW

File: src/isotools/_utils.py [hunk 1]
```diff
@@ -561,3 +561,135 @@ def cmp_dist(a, b, min_dist=3):
...
+        if x in merged_idx:
+            continue
+        for y in range(x+1, len(pos_list)):
+            if y in merged_idx:
+                continue
+            if pos_list[x] - strict_pos <= pos_list[y] <= pos_list[x] + strict_pos:
+                #print(f'{x}:{pos_list[x]} and {y}: {pos_list[y]} are close enough')
+                # keep the one with higher coverage
+                merged_idx.append(y)
+    return len(pos_list) - len(merged_idx)
```
Comment (IceFreez3r) [inline src/isotools/_utils.py:640]:
```py
sorted_pos = sorted(pos_list)
return sum([1 for i in pos_list[:-1] if pos_list[i+1] - i >= strict_pos])
```
Isn't this the same? Or is it important that high abundance positions are added first?

Comment (IceFreez3r) [reply src/isotools/_utils.py:640]: Not really sure what you're doing here.

File: src/isotools/_utils.py [hunk 2]
```diff
@@ -561,3 +561,135 @@ def cmp_dist(a, b, min_dist=3):
...
+
+
+def count_distinct_exon_chain(ec_list, strict_ec=0, strict_pos=15):
+    '''
+    :param ec_list: A list of exon chains, sorted by their abundance descendingly (output from structure_feature_cov).
+    :param strict_ec: Distance allowed between each position, except for the first/last, in two exon chains so that they cab be considered as identical.
```
Comment (IceFreez3r) [inline src/isotools/_utils.py:646]: typo "cab"

File: src/isotools/_utils.py [hunk 3]
```diff
@@ -561,3 +561,135 @@ def cmp_dist(a, b, min_dist=3):
...
+            if y in merged_idx:
+                continue
+            # if the number of exons is different, skip
+            if len(ec_list[x]) != len(ec_list[y]):
+                continue
+
+            pos_in_x = [pos for exon in ec_list[x] for pos in exon]
+            pos_in_y = [pos for exon in ec_list[y] for pos in exon]
+
+            pos_diff = [abs(x - y) for x,y in zip(pos_in_x, pos_in_y)]
```
Comment (IceFreez3r) [inline src/isotools/_utils.py:665]: Please don't use x and y here. Initial x and y are indices. These here are genomic positions

File: src/isotools/_utils.py [hunk 4]
```diff
@@ -561,3 +561,135 @@ def cmp_dist(a, b, min_dist=3):
...
+    return len(ec_list) - len(merged_idx)
+
+
+def str_var_triplet(transcripts, samples, strict_ec=0, strict_pos=15):
+    '''
+    Quantify the structure variation of transcripts in a gene across specified samples.
+    
+    :param transcripts: A list of transcript annotations of a gene obtained from isoseq[gene].transcripts.
+    :param samples: A list of sample names to specify the samples to be considered..
+    :param strict_ec: Distance allowed between each position, except for the first/last, in two exon chains so that they cab be considered as identical.
```
Comment (IceFreez3r) [inline src/isotools/_utils.py:681]: typo "cab"

File: src/isotools/plots.py [hunk 5]
```diff
@@ -357,3 +358,133 @@ def plot_rarefaction(rarefaction, total=None,  ax=None, legend=True, colors=None
     if legend:
         ax.legend()
     return ax
+
+
+def plot_str_var_number(str_var_count, group_name:'str', n_multi=10, fig_size=(12, 4), fig_title=None, **axparams):
+    '''
+    Gerenates a figure with three barplots, depicting the number of genes with a certain number of structural variations,
```
Comment (IceFreez3r) [inline src/isotools/plots.py:365]: Typo "Gerenates"

File: src/isotools/plots.py [hunk 6]
```diff
@@ -357,3 +358,133 @@ def plot_rarefaction(rarefaction, total=None,  ax=None, legend=True, colors=None
...
+        
+        axs[i].set(**{k: v for k, v in axparams.items() if k in axs[i].properties()})
+
+    if not fig_title:
+        fig_title = f'structure variation in {group_tab.columns.str.split("_").str[0].unique()[0]} samples'
+    
+    _.suptitle(fig_title, fontsize=12)
+    _.tight_layout()
+
+    return _
```
Comment (IceFreez3r) [inline src/isotools/plots.py:413]: Don't use `_` as a variable name. It's against convention

File: src/isotools/plots.py [hunk 7]
```diff
@@ -357,3 +358,133 @@ def plot_rarefaction(rarefaction, total=None,  ax=None, legend=True, colors=None
...
+def triangle_plot(str_var_tab, ax=None, colors=None, tax_title=None):
+    '''
+    Generate a triangle plot from str_var_tab.
+    Each row would be a dot in the triangle plot.
+    There can be multiple sets of three columns. Every set must be in the order of TSS, exon chain and PAS, and named as "group_feature", eg: wt_tss, wt_ec, wt_pas.
+
+    :param str_var_tab: The table of structural variation, generated by Transcriptome.str_var_calculation(count_number=False).
+    :param colors: How to color the dots in the triangle plot, either a string, a list, or a dict.
+                   If a string, all dots would be colored in the same color.
+                   If a list, there should be only one group of structural variation, and the lengthe of the list should be equal to the number of rows in str_var_tab.
```
Comment (IceFreez3r) [inline src/isotools/plots.py:425]: Typo "lengthe"

File: src/isotools/_transcriptome_io.py [hunk 8]
```diff
@@ -1462,30 +1462,50 @@ def chimeric_table(self: Transcriptome, region=None, query=None):  # , star_chim
...
-
-    with openfile(fn) as f:
+    with openfile(fn, gzip) as f:
         logger.info('writing %sgtf file to %s', "gzip compressed " if gzip else "", fn)
         for gene, transcript_ids, _ in self.iter_transcripts(genewise=True, **filter_args):
             lines = gene._to_gtf(transcript_ids=transcript_ids, source=source)
             f.write('\n'.join(('\t'.join(str(field) for field in line) for line in lines)) + '\n')
 
 
+def write_fasta(self: Transcriptome, genome_fn, fn, gzip=False, reference=False, protein=False, **filter_args):
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/_transcriptome_io.py:1490]: For 'BALANCED', 28s. Same query for gffread, 29s.

File: src/isotools/_transcriptome_stats.py [hunk 9]
```diff
@@ -16,6 +17,8 @@
 from ._utils import _filter_function
 from .splice_graph import SegmentGraph
 
+from ._utils import str_var_triplet
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/_transcriptome_stats.py:20]: There's one more above to import `ASEType`. I thought each is for different functions?

File: src/isotools/_transcriptome_stats.py [hunk 10]
```diff
@@ -406,7 +409,120 @@ def altsplice_stats(self, groups=None, weight_by_coverage=True, min_coverage=2,
...
+    if groups is None:
+        group_idxs = {'all' if len(samples) == len(self.samples) else 'selected': [self.samples.index(s) for s in samples]}
+    elif isinstance(groups, dict):
+        assert all(s in samples for s in sum(groups.values(), [])), 'not all the samples in specified groups are found'
+        group_idxs = {k: [self.samples.index(s) for s in v] for k,v in groups.items()}
+    elif isinstance(groups, list):
+        assert all(gn in self.groups() for gn in groups), 'not all specified groups are found. To customize groups, use a dict {group_name:[sample_name_list]}'
+        group_idxs = {gn: [self.samples.index(s) for s in self.groups()[gn] if s in samples] for gn in groups if any(s in samples for s in self.groups()[gn])}
+    else:
+        raise ValueError('groups must be a dict or a list of group names')
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/_transcriptome_stats.py:442]: Not really from my perspective, the checks here are different from `_check_groups`. Here the `groups` doesn't have to be all in `transcriptome.groups()`. I want to make it possible to define new groups, as long as all the samples in `groups` are identical as `samples`.

File: src/isotools/_transcriptome_stats.py [hunk 11]
```diff
@@ -406,7 +409,120 @@ def altsplice_stats(self, groups=None, weight_by_coverage=True, min_coverage=2,
...
+    if groups is None:
+        group_sns = {'all' if len(samples) == len(self.samples) else 'selected': samples}
+    elif isinstance(groups, dict):
+        assert all(s in samples for s in sum(groups.values(), [])), 'not all the samples in specified groups are found'
+        group_sns = groups
+    elif isinstance(groups, list):
+        assert all(gn in self.groups() for gn in groups), 'not all specified groups are found. To customize groups, use a dict {group_name:[sample_name_list]}'
+        group_sns = {gn: [s for s in self.groups()[gn] if s in samples] for gn in groups if any(s in samples for s in self.groups()[gn])}
+    else:
+        raise ValueError('groups must be a dict or a list of group names')
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/_transcriptome_stats.py:501]: Same reason as above. Also, here I need the sample names not sample indexes for each group.

File: src/isotools/plots.py [hunk 12]
```diff
@@ -4,6 +4,7 @@
 import seaborn as sns
 import pandas as pd
 import numpy as np
+import ternary
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/plots.py:7]: Should I add `ternary` into `requirements.txt` or `setup.cfg`, or both?

File: src/isotools/transcriptome.py [hunk 13]
```diff
@@ -261,6 +261,9 @@ def __iter__(self):
         import_sqanti_classification,
     )
 
+    # IO: proteogenomics customised database
+    from ._transcriptome_io import write_fasta
+
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/transcriptome.py:266]: They are for different functions...

File: src/isotools/_utils.py [hunk 14]
```diff
@@ -561,3 +561,135 @@ def cmp_dist(a, b, min_dist=3):
...
+
+            for s,pos_dict in tr[field].items():
+                if s in samples:
+                    for pos, n in pos_dict.items():
+                        if pos not in cov:
+                            cov[pos] = 0
+                        cov[pos] += n
+
+    # keep ones with coverage > 0
+    cov = {k: v for k, v in cov.items() if v > 0}
```
[the remark this answers is not in the document]
Comment (YalanBi) [reply src/isotools/_utils.py:612]: Not sure if I understand you. I just want to make sure only the ones with a coverage of at least one are returned.

File: src/isotools/_transcriptome_stats.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (IceFreez3r) [reply src/isotools/_transcriptome_stats.py:20]: Ah, that one is from me. Must have missed that one. Please combine all three

File: src/isotools/_transcriptome_stats.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (IceFreez3r) [reply src/isotools/_transcriptome_stats.py:442]: Maybe a new function then? Because currently you have the same code for checking groups twice

[the remark this answers is not in the document]
Comment (IceFreez3r) [reply src/isotools/_transcriptome_stats.py:442]: From below
> Also, here I need the sample names not sample indexes for each group.

Comment doesn't reflect that

File: src/isotools/plots.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (IceFreez3r) [reply src/isotools/plots.py:7]: I would guess both. Not confident though

File: src/isotools/transcriptome.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (IceFreez3r) [reply src/isotools/transcriptome.py:266]: The ones above are also not for one function

File: src/isotools/_utils.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (IceFreez3r) [reply src/isotools/_utils.py]: I understand that, but items with no coverage wouldn't be included in the dict anyways, right?
`cov[pos] = cov.get(pos, 0) + n` here `n` would need to be 0, which might not be possible?

File: src/isotools/domains.py [hunk 15]
```diff
@@ -224,7 +224,7 @@ def add_annotation_domains(self, annotation, category, id_col='uniProtId', name_
     domain_count = 0
     # clear domains of that category
     if isinstance(annotation, str):
-        anno = pd.read_csv(annotation, sep='\t')
+        anno = pd.read_csv(annotation, sep='\t', low_memory=False)
```
Comment (IceFreez3r) [inline src/isotools/domains.py:227]: Why is this needed?

Comment (YalanBi) [reply src/isotools/domains.py:227]: I got this warning/error when importing the annotation.
